Feat: Readonly wallet integration with HSM

tuliomir opened the pull request:
### Acceptance Criteria
- The headless wallet should have an integration with the HSM for a read only wallet


### Security Checklist
- [X] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

r4mmer commented on src/controllers/hsm/hsm.controller.js:
```diff
@@ -0,0 +1,149 @@
+
+  // Builds the wallet configuration object
+  const config = settings.getConfig();
+  const walletConfig = {
+    xpub: xPub,
+    password: '123',
+    pinCode: '123',
+    multisig: null,
+    connection: null,
+  };
```
suggestion: use the `getReadonlyWalletConfig` helper here.

I know it's not necessary, but since this helper is used by the normal start method it will be kept updated.

tuliomir replied:
Always good to reuse existing helper methods. Done on [519970f](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/519970fb671308993a0b28efd0cf9633440081c4).

r4mmer commented on src/services/wallets.service.js:
```diff
@@ -44,8 +54,23 @@ async function stopAllWallets() {
+ * @returns {boolean} True if this is a hardware wallet
+ */
+function isHardwareWallet(walletId) {
+  if (!initializedWallets.has(walletId)) {
+    return false;
+  }
+
+  return hardWalletIds.has(walletId);
```
```suggestion
  return initializedWallets.has(walletId) && hardWalletIds.has(walletId);
```

tuliomir replied:
Done on [519970f](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/519970fb671308993a0b28efd0cf9633440081c4)

r4mmer commented on src/controllers/hsm/hsm.controller.js:
```diff
@@ -0,0 +1,149 @@
+  const xPub = await hsmService.getXPubFromKey(connectionObj, hsmKeyName, {
+    isReadOnlyWallet: true,
+  });
+  await hsmService.hsmDisconnect();
```
question: Do we have a test to check that this method is called?

If this is not called the hsm lock will not be unlocked and any subsequent calls to HSM methods will fail.
From what I can see there are some returns that could be problematic in this route.

tuliomir replied:
Good catch! Added new tests and fixed the leaking connections on [f503f41](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/f503f41a05a401ef2a6ff21306fee4cabc21c541).

pedroferreira1 commented on src/services/hsm.service.js:
```diff
@@ -0,0 +1,203 @@
+ * @returns {Promise<Hsm>}
+ */
+async function hsmConnect() {
+  // Ensures that only one connection will be open at any given moment
```
Can't you reuse the same connection to start other wallets?

tuliomir replied:
While using the dinamo lib we currently can't. [A comment](https://github.com/HathorNetwork/rfcs/pull/67#discussion_r1462591197) was made on the design about this specific point.

pedroferreira1 commented on src/services/hsm.service.js:
```diff
@@ -0,0 +1,203 @@
+ * @returns {Promise<void>}
+ */
+async function hsmDisconnect() {
+  await hsm.disconnect();
```
Can it be disconnected somehow in the middle of an execution? How do you know that you lost the session connection?

tuliomir replied:
Yes, even though it's unlikely, an error can occur in the middle of an execution. In that case we must have good error handling because the exception must be raised and handled by the controller.

This `disconnect()` method is a no-op in case the connection has already been closed or dropped, so it's safe to call on the controller's exception handling.

tuliomir replied:
Actually, one of the starting conditions didn't have a protection for this. It was added on [6756809](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/675680910105fbd54ad684a44c88016d9db1b20c).

pedroferreira1 commented on src/services/hsm.service.js:
```diff
@@ -0,0 +1,203 @@
+  // Derivation 1: Bip Code
+  const derivationVersion = options?.version
+    ? options.version
+    : hsm.enums.VERSION_OPTIONS.BIP32_HTR_MAIN_NET;
```
Why don't you get the network in the config, so you can use the default as testnet here, in case this is the configuration?

tuliomir replied:
Added on [b332161](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/b3321613176a720d172eb4f12898e34f345ad84e)

pedroferreira1 commented on src/services/hsm.service.js:
```diff
@@ -0,0 +1,203 @@
+ * @param {Object} hsmConnection
+ * @param {string} hsmKeyName
+ * @param {Object} [options]
+ * @param {boolean} [options.isReadOnlyWallet] If true, derivation will stop at the account level
```
Can you start a non read only wallet?

tuliomir replied:
For future implementations we may want to execute the fourth derivation for calculating the change addresses on the HSM. This flag allows this alternative.

pedroferreira1 replied:
I didn't understand the reason for that, to be honest, when we want to derive until the change path and when we want to derive until the account path?

tuliomir replied:
As we've talked, I added some clarifications to the method parameters regarding which levels are derived from the inputs. Done on [2bf7108](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/2bf7108313b675fabdeb32f7a4d60bdd853d86df).

pedroferreira1 commented on src/controllers/hsm/hsm.controller.js:
```diff
@@ -0,0 +1,154 @@
+  });
+  await hsmService.hsmDisconnect();
+
+  // Builds the wallet configuration object
```
From this line below it's exactly the same as in the `/start` of the other wallets (except for the promise resolve). Can we make this refactor, so we don't duplicate code?

If we change anything from the `/start` endpoint we would need to remember to change it here as well, this is not good.

tuliomir replied:
This is a good idea. The refactor will be started on #392 and used here.

tuliomir replied:
Refactor completed, the helper method is already being used instead of the duplicated code.

pedroferreira1 reviewed (commented):
Isn't it missing an update in the `config.js.template` and in the `config.js.docker` with the new optional configs for hsm?

tuliomir commented on package-lock.json:
```diff
@@ -1126,6 +1126,11 @@
         "kuler": "^2.0.0"
       }
     },
+    "@dinamonetworks/hsm-dinamo": {
```
No, they don't. 🤩 Nice, isn't it?

tuliomir commented on src/services/wallets.service.js:
```diff
@@ -12,16 +12,26 @@ const { removeAllWalletProposals } = require('./atomic-swap.service');
+/**
+ * @type {Map<string, String>}
+ */
+const hardWalletIds = new Map();
```
Renamed on [2cce965](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/2cce965472b652cdb9dfd16e2d45f683badfc036).

tuliomir commented on src/services/wallets.service.js:
```diff
@@ -12,16 +12,26 @@ const { removeAllWalletProposals } = require('./atomic-swap.service');
 const initializedWallets = new Map();
 
+/**
+ * @type {Map<string, String>}
```
A better description was added on [2cce965](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/2cce965472b652cdb9dfd16e2d45f683badfc036)

tuliomir commented:
> Isn't it missing an update in the `config.js.template` and in the `config.js.docker` with the new optional configs for hsm?

Good point! On [ecbb17b](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/ecbb17b5c453425ea3dcbe754706888d97fb0b78) added the default configs and implemented a better user experience in case these configs were not informed when calling an HSM endpoint. Also added tests for this new error message.

pedroferreira1 commented on src/services/hsm.service.js:
```diff
@@ -0,0 +1,215 @@
+ * Derivates an HTR wallet xPub string from a given HSM xPriv key
+ * @param {Object} hsmConnection
+ * @param {string} hsmKeyName
+ * @param {Object} [options]
```
What are the options here?

tuliomir replied:
As discussed on the comment above, this parameter was removed.
